trl/trainer/kto_trainer.py

```diff
@@ -0,0 +1,963 @@
+# KTO Authors: Kawin Ethayarajh, Winnie Xu, Dan Jurafsky, and Douwe Kiela
+# Copyright 2023 The HuggingFace Team. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import inspect
+import random
+import warnings
+from collections import defaultdict
+from contextlib import nullcontext
+from copy import deepcopy
+from functools import wraps
+from typing import Any, Callable, Dict, List, Literal, Optional, Tuple, Union
+
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from accelerate.utils import is_deepspeed_available, tqdm
+from datasets import Dataset, interleave_datasets
+from torch.utils.data import DataLoader, SequentialSampler
+from transformers import (
+    AutoModelForCausalLM,
+    DataCollator,
+    PreTrainedModel,
+    PreTrainedTokenizerBase,
+    Trainer,
+    TrainingArguments,
+)
+from transformers.trainer_callback import TrainerCallback
+from transformers.trainer_utils import EvalLoopOutput, has_length
+
+from ..import_utils import is_peft_available, is_wandb_available
+from ..models import PreTrainedModelWrapper, create_reference_model
+from .dpo_trainer import DPOTrainer
```

Comment:
I think we should not re-use methods between trainers now that we don't inherit, otherwise we have the worst of both worlds: we don't inherit but still have cross dependencies. I think it's fine to copy paste and maybe have a unittest that we do it the same way in both trainers to enforce consistency. What do you think? cc @younesbelkada

Reply:
ok undoing